### PR TITLE
GPU driven determineSectors

### PR DESCRIPTION
**Pull request by larsmoa**
GPU driven implementation of determineSectors() that takes into account how much each sector will contribute to the final, visual result.

### CODE REVIEW

File: examples/src/simple.ts [hunk 1]
```diff
@@ -26,10 +37,58 @@ async function main() {
...
   const { position, target, near, far } = cadNode.suggestCameraConfig();
-  const camera = new THREE.PerspectiveCamera(75, window.innerWidth / window.innerHeight, near, far);
+  camera.near = near;
+  camera.far = 3 * far;
+  camera.fov = 75;
+  camera.aspect = window.innerWidth / window.innerHeight;
+  camera.updateProjectionMatrix();
+  camera.updateMatrixWorld();
+
+  document.addEventListener('keypress', event => {
```
Comment (dragly) [inline examples/src/simple.ts:61]: Move to helper class

File: examples/webpack.config.js [hunk 2]
```diff
@@ -230,6 +230,18 @@ module.exports = env => {
         resolve('public/'),
         resolve('dist/'),
       ],
+
+      proxy: {
```
Comment (dragly) [inline examples/webpack.config.js:234]: Unrelated change?

File: viewer/src/data/worker/WorkerPool.ts [hunk 3]
```diff
@@ -16,7 +16,15 @@ interface PooledWorker {
 }
 
 export class WorkerPool {
+  static get defaultPool(): WorkerPool {
```
Comment (dragly) [inline viewer/src/data/worker/WorkerPool.ts:19]: Unrelated change?

Comment (dragly) [reply viewer/src/data/worker/WorkerPool.ts:19]: I like it, though :)

File: viewer/src/glsl/base/updateFragmentDepth.glsl [hunk 4]
```diff
@@ -1,6 +1,4 @@
-#ifdef GL_EXT_frag_depth
-
-void updateFragmentDepth(vec3 p,mat4 projectionMatrix){
+float computeFragmentDepth(vec3 p, mat4 projectionMatrix) {
```
Comment (dragly) [inline viewer/src/glsl/base/updateFragmentDepth.glsl:1]: Unrelated change

File: viewer/src/views/threejs/cad/CadNode.ts [hunk 5]
```diff
@@ -182,16 +186,19 @@ export class CadNode extends THREE.Object3D {
...
+      (consumedSector: ConsumedSector) =>
+        currentLevelOfDetail.get(consumedSector.sectorId) !== consumedSector.levelOfDetail
     );
     const pipeline = new Subject<ThreeCameraConfig>();
     pipeline
       .pipe(
-        auditTime(100),
+        // TODO 2020-04-15 larsmoa: Reduce delay to something more sensible
+        // Temporary workaround to avoid flooding the GPU pipeline with readPixels.
+        debounceTime(150),
```
Comment (dragly) [inline viewer/src/views/threejs/cad/CadNode.ts]: Will not the change to `debounceTime` cause us to potentially never update the sectors? That is, if you continue to move the camera, debounceTime will just postpone and postpone updating forever, leaving us with nothing on screen. Why not just increase the time sent to `auditTime`?

File: viewer/src/culling/ByVisibilityGpuSectorCuller.ts [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+  coverageUtil?: OrderSectorsByVisibilityCoverage;
+};
+
+function assert(condition: boolean, message: string = 'assertion hit') {
+  // tslint:disable-next-line: no-console
+  console.assert(condition, message);
+}
+
+class TakenSectorMap {
+  get totalCost() {
```
Comment (dragly) [inline viewer/src/culling/ByVisibilityGpuSectorCuller.ts:59]: Why is this a getter, while getWantedSectorCount is a member function?

File: viewer/src/culling/ByVisibilityGpuSectorCuller.ts [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+
+    if (!changed) {
+      return takenSectors;
+    }
+    takenSectors.clear();
+    this.models.forEach(x => takenSectors.initializeScene(x.scene));
+
+    // Update wanted sectors
+    const prioritized = coverageUtil.orderSectorsByVisibility(this.camera);
+    this.lastUpdate.matrix = this.camera.matrix.clone();
```
Comment (dragly) [inline viewer/src/culling/ByVisibilityGpuSectorCuller.ts:182]: Use `this.lastupdate.matrix.copy(this.camera.matrix)` instead to avoid the extra allocation.

File: viewer/src/culling/ByVisibilityGpuSectorCuller.ts [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    let debugAccumulatedPriority = 0.0;
+    const prioritizedLength = prioritized.length;
+
+    let i = 0;
+    for (i = 0; i < prioritizedLength && takenSectors.totalCost < costLimit; i++) {
+      const x = prioritized[i];
+      takenSectors.markSectorDetailed(x.scene, x.sectorId, x.priority);
+      debugAccumulatedPriority += x.priority;
+    }
+    console.log(
```
Comment (dragly) [inline viewer/src/culling/ByVisibilityGpuSectorCuller.ts:200]: Remove logging

File: viewer/src/culling/TakenSectorTree.ts [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+      `Sector ${sectorId} must be a Simple-sector, but got ${this.sectors[sectorId].lod}`
+    );
+    this.setSectorLod(sectorId, LevelOfDetail.Detailed);
+    this.markAllDiscardedChildrenAsSimple(sectorId);
+  }
+
+  private markAllDiscardedChildrenAsSimple(sectorId: number) {
+    for (const child of this.sectors[sectorId].sector.children) {
+      if (this.getSectorLod(child.id) === LevelOfDetail.Discarded) {
+        if (child.facesFile.fileName !== null) {
```
Comment (dragly) [inline viewer/src/culling/TakenSectorTree.ts:96]: If the filename is null, should we not stop here and just use the empty sector as a simple representation rather than loading child sectors with simple data with higher number of faces?

File: viewer/src/glsl/sector/sectorCoverage.frag [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+varying lowp vec2 v_seed;
+
+
+void main() {
+    float v = rand2d(gl_FragCoord.xy + v_seed);
+    if (v >= v_coverageFactor) {
+        discard;
+    }
+
+    gl_FragColor = vec4(v_color, pow(v_position.z / v_position.w, 10.0)); // Ghetto-linearization of the depth
```
Comment (dragly) [inline viewer/src/glsl/sector/sectorCoverage.frag:16]: Why 10.0?

**Comment by dragly** [comment]
Approving, but I would prefer if the above comments are resolved.

File: examples/webpack.config.js (hunk 2 above)
Comment (larsmoa) [reply examples/webpack.config.js:234]: Yes, this was necessary in order to be able to test performance on mobile. It's cherry-picked into a separate PR (https://github.com/cognitedata/reveal/pull/391)

File: viewer/src/data/worker/WorkerPool.ts (hunk 3 above)
Comment (larsmoa) [reply viewer/src/data/worker/WorkerPool.ts:19]: Kinda unrelated, but its to support multiple models better by not flooding with N workers for each model added.

File: viewer/src/glsl/base/updateFragmentDepth.glsl (hunk 4 above)
Comment (larsmoa) [reply viewer/src/glsl/base/updateFragmentDepth.glsl:1]: Yes, cherry-picked to PR #390 - will merge this first to ensure it's not part of the final commit.

File: viewer/src/views/threejs/cad/CadNode.ts (hunk 5 above)
Comment (larsmoa) [reply viewer/src/views/threejs/cad/CadNode.ts]: Yes, it will. I considered it the lesser evil of causing stuttering while moving around. Because we load geometry for the full model at all times (just at low detail), I considered acceptable for now. I realize it's far from optimal, but I don't see a good solution right now and the upside of the new culler outbeats the disadvantage this results in.

Note! A theory I have is that we should not do `render(); readPixels();` in the same call to `determineSectors()` as this will cause the render-pipeline to flush before `readPixels` returns.

File: viewer/src/culling/ByVisibilityGpuSectorCuller.ts (hunk 6 above)
Comment (larsmoa) [reply viewer/src/culling/ByVisibilityGpuSectorCuller.ts:59]: It's maybe a stupid pattern, but I tend to use functions to indicate that there is some computational cost associated with the operation. Here getWantedSectorCount() is `O(M*N)` where M is models and N is sectors. `get totalCost` is `O(M)`.

I realize I'd have a hard time to make this argument in front of a jury and the natural follow-up-question would be "Why not clearly indicate this in the function name instead?" - and hence I will rename it to `determineWantedSectorCount()`

File: viewer/src/culling/ByVisibilityGpuSectorCuller.ts (hunk 7 above)
Comment (larsmoa) [reply viewer/src/culling/ByVisibilityGpuSectorCuller.ts:182]: Good catch, fixed.

File: viewer/src/glsl/sector/sectorCoverage.frag (hunk 10 above)
Comment (larsmoa) [reply viewer/src/glsl/sector/sectorCoverage.frag:16]: Remaints of a concept I tested, removed now.

File: viewer/src/culling/TakenSectorTree.ts (hunk 9 above)
Comment (larsmoa) [reply viewer/src/culling/TakenSectorTree.ts:98]: Good point, fixed.

File: examples/src/simple.ts (hunk 1 above)
Comment (larsmoa) [reply examples/src/simple.ts]: Created a separate example for this (short lived - will be the one and only culler when we get the new fix from J)
